lintreview/docker.py

```diff
@@ -92,44 +111,49 @@ def run(image, command, source_dir, env=None, timeout=None, name=None):
     for tool execution.
     """
     log.info('Running %s container', image)
-
-    env_args = []
-    if isinstance(env, dict):
-        for key, val in env.items():
-            env_args.extend(['-e', u'{key}={val}'.format(key=key, val=val)])
-    elif env:
-        raise ValueError('env argument should be a dict')
-
-    cmd = ['docker', 'run']
+    if not docker_base:
+        docker_base = DOCKER_BASE
+
+    run_args = {
+        'image': image,
+        'command': [str(c) for c in command],
+        'environment': env,
+        'volumes': {source_dir: {'bind': docker_base, 'mode': 'rw'}},
+        'stdout': True,
+        'stderr': include_error,
+        'detach': True,
+    }
 
     if name is not None:
-        cmd += ['--name', name]
-    else:
-        cmd.append('--rm')
-
-    cmd += [
-        '-v',
-        u'{}:{}'.format(source_dir, DOCKER_BASE)
-    ]
-    cmd += env_args
-    cmd.append(image)
-
-    # Make all the arguments into bytestr strings.
-    # to get around encoding issues.
-    cmd += [six.text_type(arg).encode('utf8') for arg in command]
-
-    log.debug('Running %s', cmd)
-    process = subprocess.Popen(
-        cmd,
-        stdin=subprocess.PIPE,
-        stdout=subprocess.PIPE,
-        stderr=subprocess.PIPE,
-        universal_newlines=True)
-
-    # Get output bytes/string
-    output, error = process.communicate()
-    output = error + output
-    log.debug('Container output was: %s', output)
+        run_args['name'] = name
+
+    if workdir:
+        run_args['working_dir'] = workdir
+
+    log.debug('Running %s', run_args)
+    client = _get_client()
+    try:
+        container = client.containers.run(**run_args)
+    except ImageNotFound:
+        err_txt = "Image not found."
+        log.exception(err_txt)
+        return err_txt
+    except APIError:
+        log.exception("API Error running container.")
+        return "API Error Running Container."
+
+    try:
+        container.wait(timeout=timeout)
+        output = b''
+        if include_error:
+            output += container.logs(stdout=False, stderr=True)
+        output += container.logs(stdout=True, stderr=False)
+    except (APIError, ReadTimeout, ConnectionError):
+        log.exception("Container.wait exception.")
+        return "Exception waiting for container to finish."
```

Comment:
This could become an error that the tool run process catches and adds to the review comments. This could be something done down the road though.

Reply:
Hmm, that's a good point, maybe I should raise a `ValueError` instead? since that's what we raise for other problems?

Reply:
A TimeoutError would work well. We could use that to re-queue the review job if it didn't already fail from a timeout error.

Reply:
I've created a new error class in the `docker` module named `TimeoutError` which theoretically could conflict with `multiprocessing.connection.TimeoutError` but as they're very different things, I think it makes sense to have a separate exception. I originally named it `DockerTimeoutError` but that seemed to stutter in the package name with `docker.DockerTimeoutError`. I'm open to changing this.

Additionally, there aren't any tests that I can see for tasks. I probably won't be able to just shove this into prod to test it until wednesday or thursday. Even then, I'm not sure how (besides with a very custom linter) to test it. Any suggestions here from your experience?